pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -304,6 +318,54 @@ def _register_ddp_hooks(self) -> None:
                 ddp_comm_wrapper=self._ddp_comm_wrapper,
             )
 
+            # Post-localSDG is only available after 1.9,
+            # and `torch.distributed.optim` package currently is not available on Windows.
+            if (
+                not _IS_WINDOWS
+                and _TORCH_GREATER_EQUAL_1_9
+                and isinstance(self._ddp_comm_state, post_localSGD.PostLocalSGDState)
+                and self.lightning_module.trainer.state.fn == TrainerFn.FITTING
+            ):
+                self._reinit_optimizers_with_post_localSGD(self._ddp_comm_state.start_localSGD_iter)
+
+    def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
```

Comment:
I believe we could maybe take the opposite approach to enable support for `DistributedOptimizer`, `ZeroRedundancyOptimizer` and sharded.

What if we patch `LightningModule.configure_optimizers` where each optimizer will be wrapped by a `PostLocalSGDOptimizer` and then we call optimizers again from the trainer.

Reply:
We could maybe move this logic to `DDPPlugin`, `DDPSpawn` setup functions, so accelerator `setup` is being called on `PostLocalSGDOptimizer` optimizers.

Reply:
> I believe we could maybe take the opposite approach to enable support for `DistributedOptimizer`, `ZeroRedundancyOptimizer` and sharded.

I am afraid distributed optimizers may not be compatible. For example, `ZeroRedundancyOptimizer` does not hold all the parameters in each shard, and does not do all-reduce. In comparison,  `PostLocalSGDOptimizer` still does all-reduce (Although we probably can develop another `PostLocalSGDOptimizer` implementation that does not do all-reduce).

It's just simple and safe to prevent the combinations in this PR.

> What if we patch `LightningModule.configure_optimizers` where each optimizer will be wrapped by a `PostLocalSGDOptimizer` and then we call optimizers again from the trainer.

I am not familiar with `LightningModule.configure_optimizers`. Are you asking what will happen if we want to use the local optimizer(s) wrapped by `PostLocalSGDOptimizer`? I don't think it will work, because I have deleted the local optimizer(s) after the rewrapping.